src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Provides a high-performance API for forward-only, read-only access to the UTF-8 encoded JSON text.
+    /// It processes the text sequentially with no caching and adheres strictly to the JSON RFC
+    /// by default (https://tools.ietf.org/html/rfc8259). When it encounters invalid JSON, it throws
+    /// a JsonReaderException with basic error information like line number and byte position on the line.
+    /// Since this type is a ref struct, it does not directly support async. However, it does provide
+    /// support for reentrancy to read incomplete data, and continue reading once more data is presented.
+    /// To be able to set max depth while reading OR allow skipping comments, create an instance of 
+    /// <see cref="JsonReaderState"/> and pass that in to the reader.
+    /// </summary>
+    public ref partial struct JsonUtf8Reader
+    {
+        private ReadOnlySpan<byte> _buffer;
+
+        private bool _isFinalBlock;
+
+        private ulong _stackFreeContainer;
+        private long _lineNumber;
+        private long _lineBytePosition;
+        private int _consumed;
+        private int _currentDepth;
+        private int _maxDepth;
+        private bool _inObject;
+        private bool _isNotPrimitive;
+        private JsonTokenType _tokenType;
+        private JsonReaderOptions _readerOptions;
+        private Stack<JsonTokenType> _stack;
+
+        private long _totalConsumed;
+        private bool _isLastSegment;
+        private readonly bool _isSingleSegment;
+        private SequencePosition _currentPosition;
+        //private SequencePosition _nextPosition; //TODO: Required for reading JSON from a ReadOnlySequeunce.
+        private ReadOnlySequence<byte> _sequence;
+
+        private bool IsLastSpan => _isFinalBlock && (_isSingleSegment || _isLastSegment);
+
+        /// <summary>
+        /// Lets the caller know which of the two 'Value' properties to read to get the 
+        /// token value. For input data within a ReadOnlySequence&lt;byte&gt; this will
+        /// always return false. For input data within a ReadOnlySequence&lt;byte&gt;, this
+        /// will only return true if the token straddles more than a single segment and
+        /// hence couldn't be represented as a span.
+        /// </summary>
+        public bool IsValueMultiSegment { get; private set; }
```

Comment:
I wonder if we could come up with a name that starts with "Value" so that it sorts next to Value and ValueSequence. Maybe "ValueIsMultisegment" is not too bad?

Reply:
```sh
~/git/dotnet/source-build/src/standard/platforms/net461$ grep [a-z]Is[A-Z] *.cs | grep get
System.Data.cs:        public bool InvokeIfReceiverIsNull { get { throw null; } set { } }
System.Xml.cs:        public bool OverrideIsNullable { get { throw null; } set { } }
mscorlib.cs:        public static bool MonitoringIsEnabled { get { throw null; } set { } }
mscorlib.cs:        public bool AssemblyIsPrivate { get { throw null; } }
```

Looks like very few.  Krzysztof's recorded opinion on this matter seems to be

> In selecting names for Boolean properties and functions, consider testing out the common uses of the API in an if--statement. Such a usage test will highlight whether the word choices and grammar of the API name (e.g., active versus passive voice, singular versus plural) make sense as English phrases.

<sub>Framework Design Guidelines (2ed), 3.6.2 (Names of Properties), annotation by Krzysztof Cwalina, p70</sub>

So the test is, which "feels" (or "sounds") better:

* `if (reader.IsValueMultiSegment) { ... }`
* `if (reader.ValueIsMultiSegment) { ... }`
* `if (reader.ValueIsMultiSegmented) { ... }`
* `if (reader.HasMultiSegmentValue) { ... }`
* `if (reader.HasMultiSegmentedValue) { ... }`
* `if (reader.HasSegmentedValue) { ... }`
* `if (reader.MultiSegmentValue) { ... }`
* `if (reader.MultiSegmentedValue) { ... }`
* `if (reader.SegmentedValue) { ... }`
* et cetera

Of that list, I think I like "HasMultiSegmentValue" the best, but arguments can be made for many (but not all) of them.

Reply:
>  I think I like "HasMultiSegmentValue" the best

Does "Segment" appear as a concept anywhere else in the API?  Maybe "HasValueSequence" so that it matches the "ValueSequence" property and is effectively just a capability test for using it?

Reply:
No "segment" doesn't show up anywhere here since it is a concept tied to ReadOnlySequence (sequence of segments), so we essentially inherit it. Still, the concept has semantic meaning here which we may want to distinguish separate from Sequence. If the input is ReadOnlySequence, the value won't always be in a ValueSequence. It will only be in ValueSequence if the value is "multi-segmented" (i.e. straddles segments).

I am fine with `HasValueSequence` except it doesn't solve @KrzysztofCwalina's original concern of having all three sort properties together. If we don't do that, I prefer to keep "multi-segment" even though it introduces a slightly new concept. Hence I would go with either what we have already, i.e. IsValueMultiSegment, or HasMultiSegmentValue.